Better Culture Recognition in Choice/Confirm Prompts

mdrichardson opened the pull request:
Fixes: #2458 

# Summary

Mostly, this brings the dotnet repo to parity with JS due to this PR: https://github.com/microsoft/botbuilder-js/pull/1121

However, I also noticed that we duplicate some code and hardcode some strings like:

[ConfirmPrompt](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder.Dialogs/Prompts/ConfirmPrompt.cs#22):

```csharp
{ Spanish, (new Choice("Sí"), new Choice("No"), new ChoiceFactoryOptions(", ", " o ", ", o ", true)) },
```

[ChoicePrompt](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder.Dialogs/Prompts/ChoicePrompt.cs#L21):

```csharp
{ Spanish, new ChoiceFactoryOptions { InlineSeparator = ", ", InlineOr = " o ", InlineOrMore = ", o ", IncludeNumbers = true } },
```

Instead, I made an [interface](https://github.com/microsoft/botbuilder-dotnet/compare/master...microsoft:cultureRecognizer#diff-021be8b8f1b0d07d1113b6dae61dd184) and [class](https://github.com/microsoft/botbuilder-dotnet/compare/master...microsoft:cultureRecognizer#diff-b817e716cb31678bef597eb46e24ddc4) to house all of the languages Confirm and Choice Prompts support. This reduces the number of places needed up update or add a language and will also make it so the prompts [dynamically support new languages that we add](https://github.com/microsoft/botbuilder-dotnet/compare/master...microsoft:cultureRecognizer#diff-e8849be75e11353fcf31c65993331c58R72).

I considered using the [PromptCultureModels](https://github.com/microsoft/botbuilder-dotnet/compare/master...microsoft:cultureRecognizer#diff-b817e716cb31678bef597eb46e24ddc4) class more in the tests, but I thought the tests might be better at catching errors if they kept hard-coded strings. **Let me know if you disagree and I should modify the tests to use the -CultureModels**

Like the JS PR, this [tests variations of locale strings](https://github.com/microsoft/botbuilder-dotnet/compare/master...microsoft:cultureRecognizer#diff-ea15b0a0329656a87a7f5b0de6765dc3R555)

**Let me know what you think of the additional Interface and Class and I can duplicate it in JS or remove it from this PR**

# Changes

## Changed

* ChoicePrompt and Tests
* ConfirmPrompt and Tests

## Added

* PromptCultureModel
* PromptCultureModels
* TestLocale (for generating test variations)

# Testing

![image](https://user-images.githubusercontent.com/40401643/64053375-f0cab600-cb36-11e9-9d95-89bdcc48a595.png)

Working in Bot:

![image](https://user-images.githubusercontent.com/40401643/64124435-e0475500-cd5b-11e9-91d7-f607c15fd7f2.png)

![image](https://user-images.githubusercontent.com/40401643/64124450-e9d0bd00-cd5b-11e9-9135-382b29494574.png)

fuselabs commented:
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.3.1)

mdrichardson commented:
**DO NOT MERGE**

Please hold off until decision made in [this issue](https://github.com/microsoft/botbuilder-dotnet/issues/2466#issuecomment-527195784).

mdrichardson commented:
@cleemullins, @sgellock (CC @mutanttech)

Regarding [this comment](https://github.com/microsoft/botbuilder-dotnet/issues/2466#issuecomment-527195784)

I had some time, so went ahead and de-coupled `Recognizers-Text.Culture` a little. See [this commit](https://github.com/microsoft/botbuilder-dotnet/pull/2463/commits/658cff004e493b84fe9f43c215354e085836e281) for more detail.

Basically, it amounts to needing to maintain:

[This](https://github.com/microsoft/botbuilder-dotnet/pull/2463/commits/658cff004e493b84fe9f43c215354e085836e281#diff-b817e716cb31678bef597eb46e24ddc4R120):

```csharp
public static string MapToNearestLanguage(string cultureCode)
{
    cultureCode = cultureCode.ToLowerInvariant();

    if (SupportedCultureCodes.All(o => o != cultureCode))
    {
        // Handle cases like EnglishOthers with cultureCode "en-*"
        var fallbackCultureCodes = SupportedCultureCodes
            .Where(o => o.EndsWith("*", StringComparison.Ordinal) &&
                        cultureCode.StartsWith(o.Split('-').First(), StringComparison.Ordinal)).ToList();

        if (fallbackCultureCodes.Count == 1)
        {
            return fallbackCultureCodes.First();
        }

        // If there is no cultureCode like "-*", map only the prefix
        // For example, "es-mx" will be mapped to "es-es"
        fallbackCultureCodes = SupportedCultureCodes
            .Where(o => cultureCode.StartsWith(o.Split('-').First(), StringComparison.Ordinal)).ToList();

        if (fallbackCultureCodes.Any())
        {
            return fallbackCultureCodes.First();
        }
    }

    return cultureCode;
}
```

...which is more or less a copy/paste from [`Recognizers-Text.Culture.MapToNearestLanguage`](https://github.com/microsoft/Recognizers-Text/blob/master/.NET/Microsoft.Recognizers.Text/Culture.cs#L64)

This will allow us to easily add additional language support for prompts, without needing to depend on `Recognizers-Text`.

fuselabs commented:
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.3.1)

mutanttech commented:
@cleemullins, @sgellock (CC @mdrichardson)
Looking forward to hear opinions from Chris and Scott soon. 

And also inputs on how long do you think this is gonna take to make into master. As I am running on kinda deadline for a Danish Bot I am working on, so this is something I am depending on to design the whole bot workflow.

If it sounds like a week or more, then kindly suggest me a short term solution for the same while getting this longer term in.

mdrichardson commented:
@mutanttech I made the updates to the Danish "or". Thanks again for the translation.

I didn't experience the same issue that you had and the confirm prompt returned the appropriate `True` when clicking "Ja":

![image](https://user-images.githubusercontent.com/40401643/64268455-fafd0380-ceec-11e9-8ba7-6dd68e322369.png)

(ignore some of the other stuff...this is just my Quick Test bot).

I'm 90% sure the issue is that you're manually setting the locale with:

```csharp
stepContext.Context.Activity.Locale = "da-DK";
```

when you should actually be setting it for the entire conversation for the user. In Emulator, you do that in Settings:

![image](https://user-images.githubusercontent.com/40401643/64268541-24b62a80-ceed-11e9-80e1-083dcb199d07.png)

In WebChat, you do that like:

```js
window.WebChat.renderWebChat(
            {
               directLine: window.WebChat.createDirectLine({
                  token: 'YOUR_DIRECT_LINE_TOKEN'
               }),
               userID: 'YOUR_USER_ID',
               username: 'Web Chat User',
               locale: 'da-DK',  // HERE
               botAvatarInitials: 'WC',
               userAvatarInitials: 'WW'
            },
            document.getElementById('webchat')
         );
```

fuselabs commented:
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.3.1)

cleemullins commented on libraries/Microsoft.Bot.Builder.Dialogs/Prompts/IPromptCultureModel.cs:
```diff
@@ -0,0 +1,56 @@
+    /// <summary>
+    /// Culture model used in Choice and Confirm Prompts.
+    /// </summary>
+    public class IPromptCultureModel
```
This would be better as a standard class ("PromptCultureModel"). When it's an interface like this, we can't add to it in the future w/o breaking compat.

mdrichardson replied:
@cleemullins Understood. Chalk it up to me still being C#-dumb. Is this the direction we want to go with this, then: de-coupling from Text-Recognizers a little and using our own Culture Models? If so, I'll make this change tomorrow and do a PR for the JS side.

mdrichardson replied:
@cleemullins Went ahead and made the change. I will interpret a Merge of this PR as approval of the changes and a request to create a parity PR for the JS side.

mutanttech commented:
@mdrichardson ,
Thank you for updating the code and doing is real quick.
And also thanks for pointing out that code. I have managed to get it to work.
And instead of getting the whole SDK or event Dialogs package, I have picked the three files that you have created in culture recognizer branch with same namespace, just renamed them to ChoicePromptCustom and ConfirmPromptCustom. And in Bot Dialog add them. It seems to work as expected. I will upgrade to 4.6 when it arrives and get rid of these additional files.
Let me know if you think any other thing might be required in addition to those files.

fuselabs commented:
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.3.1)

mdrichardson commented:
@cleemullins This is a long thread, so I'll re-summarize the PR. Changes are made in both Confirm and Choice prompts.

1. Gives users ability to set their own `Dictionary` of locale settings (`CustomLocaleOptions`)
2. Creates two additional classes to contain constants of supported languages
3. Uses `Recognizers-Text.Culture` for locale string ("en-US")
4. Uses `Recognizers-Text.Culture` to appropriately map user and SDK-provided locale string to nearest supported ("en-us", "en-US", "en" `=>` "en-US")
5. Dynamically adds SDK-Supported languages to both Confirm and Choice Prompt so in the future, we only need to add supported languages to one file
5. Tests for all of the above

![image](https://user-images.githubusercontent.com/40401643/65173413-8c976580-da03-11e9-8fc9-404e4c4064f7.png)

fuselabs commented:
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.3.1)

cleemullins commented on libraries/Microsoft.Bot.Builder.Dialogs/Prompts/ChoicePrompt.cs:
```diff
@@ -50,6 +38,18 @@ public ChoicePrompt(string dialogId, PromptValidator<FoundChoice> validator = nu
+        ///         { "en-US", new ChoiceFactoryOptions {...} }
+        /// </remarks>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.DocumentationRules", "SA1629:Documentation text should end with a period", Justification = "Doesn't like `Dictionary<string`")]
+        public static Dictionary<string, ChoiceFactoryOptions> CustomLocaleOptions { get; set; }
```
This has implications beyond the obvious. By making this dictionary public, we cannot make any threading and concurrency assumptions about it. I would *far* prefer the dictionary itself to remain private, and only allow accessors to get/set. That way we can be correct with regards to locks and concurrency.

cleemullins replied:
I should add that your code comment helps, but I would like to see something more clean.

mdrichardson replied:
Ah yeah. I should know better than that. Will fix both of those today.

mdrichardson replied:
@cleemullins Changes applied, but API Compatibility Validation is throwing errors because adding an optional parameter to the existing ctor vs. adding an override. Do we want to manually bypass this or add the override?

fuselabs commented:
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.3.1)

cleemullins commented on libraries/Microsoft.Bot.Builder.Dialogs/Prompts/PromptCultureModels.cs:
```diff
@@ -0,0 +1,122 @@
+﻿using System;
```
// Copyright

cleemullins commented on libraries/Microsoft.Bot.Builder.Dialogs/Prompts/PromptCultureModel.cs:
```diff
@@ -0,0 +1,56 @@
+﻿namespace Microsoft.Bot.Builder.Dialogs.Prompts
```
// Copyright

fuselabs commented:
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.5.3)

mdrichardson commented:
@cleemullins I don't think I'll be able to code review/sync in person until next week (out sick). Happy to then, if you want.

Also, I'm not sure why MacLinux build is failing. Is that something I can fix? It looks like the error message points to the VM, more so than the repo/SDK.

fuselabs commented:
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.5.3)

tomlm commented:
This looks good to me, but I would like 
* the Conflicting merge to be inspected to make sure we aren't losing whatever changes happened in that table
* @stevenic to review this.

mdrichardson commented:
@tomlm Conflict resolved. We don't lose the table; it's covered [here](https://github.com/microsoft/botbuilder-dotnet/pull/2463/files#diff-b817e716cb31678bef597eb46e24ddc4R112)

fuselabs commented:
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.5.3)

fuselabs commented:
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.5.3)

cleemullins commented:
I believe this gives us partiy in both C# and JS at this point. I think what's here is reasonable. I'll get one more set of eyes on it, and then we can merge.

garypretty commented on libraries/Microsoft.Bot.Builder.Dialogs/Prompts/ConfirmPrompt.cs:
```diff
@@ -184,10 +208,10 @@ protected override Task<PromptRecognizerResult<bool>> OnRecognizeAsync(ITurnCont
-            var culture = activity.Locale ?? DefaultLocale;
-            if (string.IsNullOrEmpty(culture) || !ChoiceDefaults.ContainsKey(culture))
+            var culture = MapToNearestLanguage(activity.Locale ?? DefaultLocale ?? English.Locale);
+            if (string.IsNullOrEmpty(culture) || !_choiceDefaults.ContainsKey(culture))
```
Will culture ever be null here anymore?

Stevenic commented on libraries/Microsoft.Bot.Builder.Dialogs/Prompts/ChoicePrompt.cs:
```diff
@@ -1,167 +1,189 @@
-    /// </summary>
-    public class ChoicePrompt : Prompt<FoundChoice>
-    {
-        private static readonly Dictionary<string, ChoiceFactoryOptions> DefaultChoiceOptions = new Dictionary<string, ChoiceFactoryOptions>(StringComparer.OrdinalIgnoreCase)
```
This is a breaking change...
